### PR TITLE
[Java SDK] Implementation of Java SDK

### PR DESCRIPTION
**Pull request by kenvifire**
This feature is still working in progress.
Tasks:
- [x] Set up wasmedge-jni project
- [x] Set up wasmedge-java project
- [ ] Wrap c apis
    - [ ] WasmEdge Basics
       -  [x] Version
       -  [x] Logging Settings
       -  [x] Value Types
       -  [x] Strings
       -  [x] Results
       -  [ ] Contexts
           - [x] ConfigureContext
           - [x] FunctionTypeContext
           - [x] ASTModuleContext
           - [x] ExecutorContext
           - [x] ExportTypeContext
           - [x] GlobalTypeContext
           - [x] ImportTypeCOntext
           - [x] LoaderContext
           - [x] MemoryTypeContext
           - [x] LoaderContext
           - [x] StoreContext
           - [x] StatisticsContext.
           - [x] TableTypeContext
           - [x] ValidatorContext
    - [ ] WasmEdge VM
      - [x] VM Creation
      - [ ]  Preregistrations
      - [x]  Host Module Registrations
      - [x]  Wasm Registration and Executions
      - [x] Instance Tracing
    - [ ] WasmEdge Runtime
       - [x] Loader
       - [x] Validator
       - [x] Executor
       - [x] AST Module
       - [x] Store
       - [x] Instances
       - [ ] Host Functions
     - [ ] AOT compiler
        - [x] Compiler options    
      - [ ] Async API
- [ ] Documentation
   - [x] Readme, build and install instructions 
   - [ ] Java API docs
   - [ ]  Step by step examples

### CODE REVIEW

**Comment by juntao** [comment]
Great. Thanks! Do we have an example that shows how to run a Wasm program from Java such as these?

https://github.com/WasmEdge/WasmEdge/blob/master/docs/run.md

**Comment by hydai** [comment]
> Great. Thanks! Do we have an example that shows how to run a Wasm program from Java such as these?
> 
> https://github.com/WasmEdge/WasmEdge/blob/master/docs/run.md

This PR is a basic setup for Java SDK. And it only contains the getVersion feature. I don't think it can run a Wasm program at the current stage.

**Comment by kenvifire** [comment]
@juntao @hydai This PR is still WIP. I'll let you guys know once it's completed done.

**Comment by kenvifire** [comment]
@juntao @hydai Most part are done, please help review this MR. I'll continue working on the rest part.

**Comment by hangedfish** [comment]
Cool, is there a preview in the Maven repository? I can't wait to try it.

**Review by hydai — COMMENTED** [review]
Since there are several changes made in the nearest release, could you provide a CI workflow to build and test this Java SDK?

**Comment by hydai** [comment]
This is the rust binding workflow: https://github.com/WasmEdge/WasmEdge/blob/master/.github/workflows/bindings-rust.yml
Basically, it will contain the building process and the testing process at least in the Linux environment.

**Comment by kenvifire** [comment]
@hydai Updated API to the latest and also added CI workflow following your suggestion. Can you please take a look?

**Comment by hydai** [comment]
Hi @kenvifire 
I found there are some warnings in https://github.com/WasmEdge/WasmEdge/actions/runs/3258891835/jobs/5351315117#step:8:1

These warnings may be big issues if the users build it with `-Werror`, could you please fix them?

**Comment by kenvifire** [comment]
@hydai Fixed warnings.

**Comment by kenvifire** [comment]
@hydai fixed warnings.

**Comment by hydai** [comment]
Hi @kenvifire 
I and @dannypsnl will start to review the remaining implementations. It will take some time, but I think we can finish it this week.

File: bindings/java/wasmedge-jni/lib/common.c
Comment (dannypsnl) [inline bindings/java/wasmedge-jni/lib/common.c:118]: The `gradle test` failed currently, I think that might relate to the error from C++ not get converted to Java exception properly, then JUnit cannot handle the problem.

File: bindings/java/wasmedge-jni/lib/AstModuleContext.c
Comment (dannypsnl) [inline bindings/java/wasmedge-jni/lib/AstModuleContext.c:1]: and for all `*.c` in JNI, I think you can format them with clang-format, following the standard as `*.cpp` code in WasmEdge.

File: bindings/java/wasmedge-jni/lib/common.c
Comment (dannypsnl) [reply bindings/java/wasmedge-jni/lib/common.c:118]: I separate them into two sets
### success
- ASTModuleContextTest 
- ConfigureContextTest 
- FunctionTypeContextTest
- GlobalInstanceContextTest 
- GlobalTypeContextTest 
- ImportTypeContextTest 
- LoaderContextTest 
- MemoryInstanceTest 
- MemoryTypeContextTest
- ExportTypeContextTest
- StatisticsContextTest
- TableInstanceTest
- TableTypeContextTest
- ValidatorContextTest
- ValueTest
- WasmEdgeTest
- CompilerContextTest
### failed
- ModuleInstanceContextTest 
- ExecutorContextTest 
- BaseTest 
- StoreContextTest 
- WasmEdgeVMTest

Comment (dannypsnl) [reply bindings/java/wasmedge-jni/lib/common.c:118]: Here it's
```
> Task :test FAILED
Assertion failed: (This == nullptr), function Executor, file executor.h, line 95.
```

Comment (kenvifire) [reply bindings/java/wasmedge-jni/lib/common.c:118]: All the test cases passed both in [ci workflow](://github.com/WasmEdge/WasmEdge/actions/runs/3277515842/jobs/5394895796) and my local Ubuntu dev box. Can you provide your evn configuration and build steps for me to reproduce this issue?

Comment (dannypsnl) [reply bindings/java/wasmedge-jni/lib/common.c:118]: Yes, my environment is quite unstable right now, just ignore this for now :). I will try to figure out the root cause.

File: bindings/java/wasmedge-java/build.gradle
Comment (dannypsnl) [inline bindings/java/wasmedge-java/build.gradle:29]: repeated line 2? Is this expected?

Comment (dannypsnl) [reply bindings/java/wasmedge-java/build.gradle:32]: I have already ensured the latest version can work, I will approve once this is resolved.

**Review by dannypsnl — APPROVED** [review]

**Comment by dannypsnl** [comment]
@kenvifire You would need to signoff every commit, maybe an alias git command to `commit -s` would help.

File: bindings/java/wasmedge-jni/README.md
Comment (kenvifire) [reply bindings/java/wasmedge-jni/README.md:1]: > Maybe a link to wasmedge build steps, like wasmedge-java did
> 
> ### example
> * Install wasmedge first [[link]]
> * Go to `WasmEdge/bindings/java/wasmedge-jni`
> * ...

Already had this part in README under wasmedge-java.

**Comment by kenvifire** [comment]
@dannypsnl Can we merge this PR first? So that we can have others working on this feature at the same time.

**Review by dannypsnl — APPROVED** [review]

**Comment by kenvifire** [comment]
@hydai I've got one approval, what's the next step for merging this pull request?

**Comment by hydai** [comment]
Thank you @kenvifire 
Nice work.

**Comment by kenvifire** [comment]
> Thank you @kenvifire 
> 
> Nice work.

Cheers! By the easy, I'm still working on other features, ci and some code refactoring. Is there any way I can track my work? Or can I just create some issues to track them?

**Comment by hydai** [comment]
Hi @kenvifire 
It would be better to have an issue recording your milestones.

**Comment by kenvifire** [comment]
> Hi @kenvifire It would be better to have an issue recording your milestones.

Created an issue: https://github.com/WasmEdge/WasmEdge/issues/2025. Can you please assign it to me? Thanks.

**Pull request merged**
